datastore/query.go

```diff
@@ -702,21 +812,21 @@ func (c *Client) GetAll(ctx context.Context, q *Query, dst interface{}) (keys []
 					// an ErrFieldMismatch is returned.
 					errFieldMismatch = err
 				} else {
-					return keys, err
+					return res, err
 				}
 			}
 			if mat != multiArgTypeStructPtr {
 				ev = ev.Elem()
 			}
 			dv.Set(reflect.Append(dv, ev))
 		}
-		keys = append(keys, k)
+		res.Keys = append(res.Keys, k)
 	}
-	return keys, errFieldMismatch
+	return res, errFieldMismatch
 }
 
-// Run runs the given query in the given context.
-func (c *Client) Run(ctx context.Context, q *Query) *Iterator {
+// Run runs the given query in the given context with the provided options
+func (c *Client) Run(ctx context.Context, q *Query, opts ...RunOption) *Iterator {
```

Comment:
kind of related - here we are just adding `...RunOptions` as a new parameter rather than creating a new method `RunWithOptions`. Why do we need to create a new method `GetAllWithOptions` if we could just add them to `Get`?

Reply:
Run method returns a struct pointer called Iterator. So, I'm able to add explainmetrics as a field to the existing struct.
A struct is somewhat similar to a class in Java. Here, pointer to object is being returned. New fields can be introduced to existing object.


In case of GetAll, this is the method signature:
```
func (c *Client) GetAll(ctx context.Context, q *Query, dst interface{}) (keys []*Key, err error)
```
This cannot be changed to 
```
func (c *Client) GetAll(ctx context.Context, q *Query, dst interface{}) (keys []*Key, err error, em *ExplainMetrics)
```
Because this will break client code:
```
dst := []*SQChild{}
_ , err := client.GetAll(ctx, query, &dst)
```
With error:  `assignment mismatch: 2 variables but client.GetAll returns 3 values`

Reply:
got it, thank you for the detailed explanation :)